19490 Upgrade select2 to 3.5.x

myabc opened the pull request:
http://community.openproject.org/work_packages/19490

myabc committed:
Bump select2 to latest in 3.5.x series

User Story # 19490

Signed-off-by: Alex Coles <alex@alexbcoles.com>

Commit:
update select2 style for new version

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://myabc-feature-19490-select2-3-5-upgrade.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-19490-select2-3-5-upgrade.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
Will be taking this... Lets see how good a can kick frontend ^^

NobodysNightmare committed:
remove superfluous resolution

NobodysNightmare committed:
resolve issues with selection formatting

apparently in case of "nothing" an empty array is passed to "formatSelection"
an array does neither understand "name" nor "project"

NobodysNightmare committed:
fix some XSS vulnerabilities

those vulnerabilities were introduced with Select2 3.5.2, which
apparently allows markup in the result of formatSelection.

I could not reproduce the vulnerabilities with the previously used Select2 3.3.2

NobodysNightmare committed:
enter values into select2 using capybara-select2

The hand written way was broken when trying to provoke XSS

NobodysNightmare committed:
fix the XSS sample

NobodysNightmare commented:
I expect the latest commit to succeed.

_Fingers crossed_

myabc commented on features/step_definitions/project_member_steps.rb:
```diff
@@ -84,24 +84,12 @@ def select_principal(principal)
-def select_within_select2(to_select, scope)
-  tries = 3
-  begin
-    enter_name_with_select2(to_select, scope)
```
@NobodysNightmare you should be able to delete `enter_name_with_select2` too.

NobodysNightmare replied:
That's what I thought too, but nope: https://github.com/opf/openproject/pull/2835/files#diff-48cc8419001b55eb4fd45894e219d88aR132

NobodysNightmare committed:
extract method for common behaviour

NobodysNightmare commented on app/assets/javascripts/timelines_select_boxes.js:
```diff
@@ -29,6 +29,10 @@
 //requires 'autocompleter'
 
 jQuery(document).ready(function($) {
+  var formatSelection = function(item) {
```
RFC: I saw similar code without the `var` keyword.
If I understand it correctly, the `var` keyword is what I want, because I want to limit the scope of this method. It shall not be available elsewhere.

Still not sure whether I really need that keyword in this specific case or not.

floriank replied:
You do want `var` when declaring a new variable, by doing it you are binding that variable `formatSelection` to the anonymous `function($) { // [...]` above. 

If you had not done that, the variable would have been bound to the `global` scope (in a browser, that would be the `window` object). 

You could have also gone with the shorthand for declaring a function:

``` javascript
function formatSelection(item) { // [...]
```

Personally, I find the declaration using `var` better, as it emphasizes on functions being first-class and declares the  scope in an explicit manner.
